master/internal/task/allocation.go

```diff
@@ -4,137 +4,165 @@ import (
-		// Encapsulates logic of rendezvousing containers of the currently
-		// allocated task. If there is no current task, or it is unallocated, it is nil.
-		rendezvous *rendezvous
-		// proxy state
-		proxies []string
-		// records whether the allocation has completed any all gathers.
-		allGatherFinished bool
-
-		logCtx          detLogger.Context
-		restored        bool
-		portsRegistered bool
-	}
-
-	// MarkResourcesDaemon marks the given reservation as a daemon. In the event of a normal exit,
-	// the allocation will not wait for it to exit on its own and instead will kill it and instead
-	// await it's hopefully quick termination.
-	MarkResourcesDaemon struct {
-		AllocationID model.AllocationID
-		ResourcesID  sproto.ResourcesID
-	}
-	// AllocationExited summarizes the exit status of an allocation.
-	AllocationExited struct {
-		// userRequestedStop is when a container unexpectedly exits with 0.
-		UserRequestedStop bool
-		Err               error
-		FinalState        AllocationState
-	}
-	// AllocationState requests allocation state. A copy is filled and returned.
-	AllocationState struct {
-		State     model.AllocationState
-		Resources map[sproto.ResourcesID]sproto.ResourcesSummary
-		Ready     bool
-
-		Addresses  map[sproto.ResourcesID][]cproto.Address
-		Containers map[sproto.ResourcesID][]cproto.Container
-	}
-	// AllocationReady marks an allocation as ready.
-	AllocationReady struct{}
-	// AllocationWaiting marks an allocation as waiting.
-	AllocationWaiting struct {
-		Message string
-	}
-	// SetAllocationProxyAddress manually sets the allocation proxy address.
-	SetAllocationProxyAddress struct {
-		ProxyAddress string
-	}
-	// IsAllocationRestoring asks the allocation if it is in the middle of a restore.
-	IsAllocationRestoring struct{}
-)
+const killCooldown = 15 * time.Second
+
+// AllocationSignal is an interface for signals that can be sent to an allocation.
+type AllocationSignal string
 
 const (
-	killCooldown       = 15 * time.Second
-	okExitMessage      = "allocation exited successfully"
-	missingExitMessage = ""
+	// KillAllocation is the signal to kill an allocation; analogous to in SIGKILL.
+	KillAllocation AllocationSignal = "kill"
+	// TerminateAllocation is the signal to kill an allocation; analogous to in SIGTERM.
+	TerminateAllocation AllocationSignal = "terminate"
 )
 
-// NewAllocation returns a new allocation, which tracks allocation state in a fairly generic way.
-func NewAllocation(
+// AllocationState requests allocation state. A copy is filled and returned.
+type AllocationState struct {
+	State     model.AllocationState
+	Resources map[sproto.ResourcesID]sproto.ResourcesSummary
+	Ready     bool
+
+	Addresses  map[sproto.ResourcesID][]cproto.Address
+	Containers map[sproto.ResourcesID][]cproto.Container
+}
+
+// FirstContainer returns the first container in the allocation state.
+func (a AllocationState) FirstContainer() *cproto.Container {
+	for _, cs := range a.Containers {
```

Comment:
[Iteration order for maps isn't guaranteed](https://go.dev/blog/maps); is this meant to be a consistent order? Or is it meant to be the first container of an arbitrary map element? I guess I'm not really sure what these `ResourcesID`s are, so maybe there's really only ever going to be one.

Edit: wow, this just saw that this is pre-existing. Maybe a bug that should be fixed later?

Reply:
Yeah, I think it is just to report telemetry. I'll rename it to `Single` in the orm-style for 'limit 1'.